20230209 dnsdist

peterthomassen opened the pull request:
To-do:
- [x] settle on value for TCP Fast Open queue size
- [x] figure out how to REFUSE queries for catalog.internal.
- [x] ensure that other ports (webserver) are not exposed, especially on IPv6 (Docker has an issue there, may need iptables)
- [x] Deal with `Limit of simultaneous TCP connections reached - raise max-tcp-connections`
- [x] Remove query cache?

peterthomassen committed:
fix(dnsdist): adjust ACL for console access

Problem uncovered via 3d7062ddb0e3a16815888b061b29fba62e09368d

peterthomassen committed:
feat(): shift port 53 responsiblity directly to ns

peterthomassen committed:
chore(dnsdist): make TLS port explicit

peterthomassen committed:
feat(ns): use SO_REUSEPORT

peterthomassen committed:
feat(dnsdist): use SO_REUSEPORT

peterthomassen committed:
feat(ns): use as many threads as there are processing units

peterthomassen commented:
TCP Fast Open queue size: People in DNS OARC Mattermost don't know and wonder the same, but Mongo uses 1024, so I went for that.

peterthomassen commented:
![image](https://user-images.githubusercontent.com/4242683/217840471-f270e55d-bb07-451d-9249-72387bd233a5.png)

That is, if we expose pdns directly, we also serve catalog.internal. - I don't think that's problematic, as that zone does not have DNSSEC (i.e. no enumeration possible), and the qnames therein are hashes of domain names we managed, with PTR records pointing to that name. I don't see any room for abuse here, so let's just expose it.

peterthomassen committed:
chore(): remove obsolete variable and network

nils-wisiol commented on docker-compose.yml:
```diff
@@ -37,6 +37,7 @@ services:
         ipv4_address: 10.16.3.3
     sysctls:
       net.ipv4.icmp_echo_ignore_broadcasts: 0
+      net.ipv4.tcp_fastopen: 3
```
Consider using `2` (enable for incoming connections only) instead of `3` (enable for inbound and outbound connections), or confirm that using fast open for client connections does not have any security implications

peterthomassen replied:
This is bitwise, like chmod. 1 is default, and what I did is turning on the second bit, so the value now is 3. Would we have motivation to turn off previously allowed outgoing TFO connections if an application desires to use one?

Having said that, @vdukhovni pointed out in DNS OARC Mattermost that using TFO may cause session breakage due to a kernel bug. The [fix](https://github.com/torvalds/linux/commit/c7aab4f17021b636a0ee75bcf28e06fb7c94ab48) is available from Linux 5.18 on and does not seem to have been backported to Ubuntu 22.04 LTS. (It's available in Ubuntu 22.10.) Given that our TCP traffic ranges between 0.3% and 1.0% depending on circumstances, the load impact will be minimal.

I'd thus propose to drop TFO for now. Ok with you?

peterthomassen replied:
done

nils-wisiol commented on ns/conf/pdns.conf.var:
```diff
@@ -1,6 +1,7 @@
 loglevel=5
 api=yes
 api-key=${DESEC_NS_APIKEY}
+cache-ttl=60  # as advised in https://doc.powerdns.com/authoritative/performance.html
```
The referenced page says

> Not all backends may benefit from the packet cache. If your backend is memory based and does not lead to context switches, the packet cache may actually hurt performance.

Is that applicable to our situation?

peterthomassen replied:
Good point! In fact, our backend is in-memory and has [practically no context switches](http://www.lmdb.tech/bench/inmem/scaling.html).

I pushed a commit disabling the cache. -- TODO fix base commit's title

peterthomassen replied:
done

nils-wisiol commented:
Suggestion to validate the effectiveness of these changes with before/after performance test.

Before deployment some attention needs to be given to DoT/DoQ, as they have no (or little) coverage by the tests.

nils-wisiol reviewed (changes requested):
The `cache-ttl` setting needs some discussion before this can be merged

peterthomassen commented:
> Suggestion to validate the effectiveness of these changes with before/after performance test.

Yes. As I think it's reasonable to assume that none of the changes will cause performance degradation, I think it's enough to compare the current state with the state after this PR (i.e. not commit by commit).

> Before deployment some attention needs to be given to DoT/DoQ, as they have no (or little) coverage by the tests.

Yes. I changed the networking quite a bit and am surprised that the tests passed (I would have bet I had made a mistake). But the "Run e2e2 Tests" stage took 2:50h now whereas it took 0:50h before. That's peculiar.

Unfortunately, I haven't had time to look into these things, and I likely won't have before Feb 20.

nils-wisiol committed:
feat(): adds dnsperf benchmark tool

# Generate dnsperf input file from a frontend server traffic dump dump.pcap
tcpdump -nr dump.pcap "dst port 53" | grep -oE '\s[A-Z]+\?\s.*' | awk '{print $2 " " substr($1, 0, length($1) - 1)}' | gzip > dnsperf.in.gz

# Run
docker-compose up -d ns && zcat dnsperf.in.gz | docker-compose -f docker-compose.yml -f docker-compose.dnsperf.yml run dnsperf dnsperf -s ns

nils-wisiol commented:
I've run performance tests against main and this branch, which some unexpected results:

Performance of this branch:

	Queries sent:         2359455
	Queries completed:    2359455 (100.00%)
	Queries lost:         0 (0.00%)

	Response codes:       NOERROR 2216757 (93.95%), SERVFAIL 188 (0.01%), NXDOMAIN 89477 (3.79%), REFUSED 53033 (2.25%)
	Average packet size:  request 31, response 59
	Run time (s):         47.778935
	Queries per second:   49382.745764

	Average Latency (s):  0.002018 (min 0.000072, max 0.011304)
	Latency StdDev (s):   0.000532


Performance of main:

	Queries sent:         2359455
	Queries completed:    2359455 (100.00%)
	Queries lost:         0 (0.00%)

	Response codes:       NOERROR 2216757 (93.95%), SERVFAIL 188 (0.01%), NXDOMAIN 89477 (3.79%), REFUSED 53033 (2.25%)
	Average packet size:  request 31, response 59
	Run time (s):         24.540079
	Queries per second:   96147.000994

	Average Latency (s):  0.001027 (min 0.000018, max 0.007213)
	Latency StdDev (s):   0.000244

Tests are based on real-world query data and run on my laptop.

nils-wisiol committed:
fix(ns): adjust code to match code comment

peterthomassen committed:
feat(ns): increase max-tcp-connections from 20 to 4096

peterthomassen committed:
chore(ns): version bumps

peterthomassen committed:
feat(ns): disable query cache

peterthomassen committed:
feat(ns): raise EDNS buffer size to 1400 bytes

peterthomassen committed:
chore(replicator): version bumps

peterthomassen commented:
I ran tests with dnsperf, and found all f1d531daca1cb26e90e15a5c4d97808b696507b3, 5889fc3014ee74e5613de63fc5788c954feacd10, 8f02e6b7b679b23712bfb5c5bb46ff08f353f88d, 4ddf468fa3d51d99fafec0ae24b6cd514597b4ab to have a positive impact on performance (or neutral, but expected to provide a benefit for other traffic patterns).

Disabling the packet cache (`cache-ttl=0`), although advised in https://doc.powerdns.com/authoritative/performance.html#packet-cache, decreased performance by about 100%.

Turning of distributor multithreading (`distributor-threads=1`) was found to come with a performance penalty of about 10%.

ce9a437f2154e591a434adc8d0434398ec835678 was upon @hs-viktor's request.

My assessment is that this configuration is now sufficiently tested for deployment.

peterthomassen committed:
chore(openvpn): bump base image version, closes #63

peterthomassen committed:
chore(dnsdist): upgrade to 1.8

peterthomassen committed:
feat(dnsproxy): switch to official Docker image and bump version

nils-wisiol reviewed (approved):
lgtm
